### PR TITLE
(BOLT-163) Configurable connection timeout

### PR DESCRIPTION
**Pull request by Iristyle**
Supercedes #81 

Needs a few more tests:

- [x] bad hostname
- [x] port that is open but cannot be connected to given specific protocol (i.e. try WinRM connection to SSH port)
- [x] port that is not open

Needs some extra info on why WinRM is now configured as such.

### CODE REVIEW

File: spec/bolt/node/ssh_spec.rb [hunk 1]
```diff
@@ -98,6 +106,35 @@
         ssh.connect
       end
     end
+
+    it "takes < 2 seconds if connection fails to non-SSH port", ssh: true do
+      bogus_port = 5555
+      Process.spawn("nc -l #{bogus_port}")
```
Comment (Iristyle) [inline spec/bolt/node/ssh_spec.rb]: This is gross.. probably want to try to use something cross-platform, like `TCPServer.new`.  Also, can remove `ssh: true`

File: spec/bolt/node/winrm_spec.rb [hunk 2]
```diff
@@ -35,25 +35,26 @@ def stub_winrm_to_raise(klass, message)
...
-                        'CONNECT_ERROR',
-                        /Failed to connect to/) do
-        winrm.connect
+      # timeout will never occur given default 1 second connection timeout
+      Timeout.timeout(2) do
+        expect_node_error(Bolt::Node::ConnectError,
+                          'CONNECT_ERROR',
+                          /Failed to connect to/) do
+          winrm.connect
+        end
```
Comment (Iristyle) [inline spec/bolt/node/winrm_spec.rb]: Should probably add a `winrm` test that validates that it gives up after the given `timeout` seconds, like the SSH tests above using netcat.

File: lib/bolt/cli.rb [hunk 3]
```diff
@@ -140,6 +140,10 @@ def create_option_parser(results)
                 "(Optional, defaults to 100)") do |concurrency|
           results[:concurrency] = concurrency
         end
+        opts.on('-t', '--timeout TIMEOUT', Integer,
```
Comment (Iristyle) [inline lib/bolt/cli.rb]: Should this be called `timeout` or `connection_timeout`?  I would expect we're able to eventually implement some sort of `execution_timeout` (on the WinRM side the functionality is already there, just not exposed)

Comment (joshcooper) [reply lib/bolt/cli.rb]: 👍  to `connection_timeout` as it's fairly common to also expose `read_timeout`

Comment (MikaelSmith) [reply lib/bolt/cli.rb]: Probably agree with `connection_timeout`. How often does this need to be used?

File: lib/bolt/node/winrm.rb [hunk 4]
```diff
@@ -12,9 +12,17 @@ def initialize(host, port, user, password, shell: :powershell, **kwargs)
...
     def connect
-      @connection = ::WinRM::Connection.new(endpoint: @endpoint,
-                                            user: @user,
-                                            password: @password)
+      options = { endpoint: @endpoint,
+                  user: @user,
+                  password: @password,
+                  retry_limit: 1 }
+
+      if @timeout
```
Comment (MikaelSmith) [inline lib/bolt/node/winrm.rb]: Is this both connection and execution timeouts? Why are we changing from the default of `receive_timeout` being 10 seconds longer than `operation_timeout`?

Comment (MikaelSmith) [reply lib/bolt/node/winrm.rb]: Also, why are we changing the `retry_limit`?

Comment (Iristyle) [reply lib/bolt/node/winrm.rb]: I need to play a bit more with `receive_timeout` to see if it actually influences `operation_timeout` in a meaningful way. The initial defaults seemed a bit arbitrary based on observed behavior.

`retry_limit` is what causes the `timeout` to be completely ineffectual - i.e. what happens by default (with `retry_limit` of 3) is:

* fails to connect
* sleep for 10 seconds (default retry sleep)
* retry
* fails to connect
* sleep for 10 seconds
* retry
* fails to connect
* retry limit hit, fail

So for a connection failure, by default, it takes 20+ seconds before it fails, which doesn't make a lot of sense when the intent is to wait up to 1 second for a fail.

There are some notes in the commit about this, and I'll add more detail when this is finished.

File: lib/bolt/cli.rb (hunk 3 above)
Comment (Iristyle) [reply lib/bolt/cli.rb]: I think we should set low sensible defaults (1 second seems reasonable to me), making `connection_timeout` rarely used in cases where network connectivity is a problem.

**Comment by nicklewis** [comment]
Updated the tests to not shell out to `nc`, and renamed the option to `--connect-timeout`.

I'm still having some trouble with WinRM in the case where the port is entirely unresponsive. The timeout doesn't seem to be getting hit there...

**Comment by nicklewis** [comment]
Okay this seems to be working now. The default connect timeout wasn't being honored but is now, and I changed the way we apply the timeout to WinRM so that it actually works.

**Comment by nicklewis** [comment]
Tests pass! I think this is ready to go.

File: lib/bolt/cli.rb [hunk 5]
```diff
@@ -144,6 +144,10 @@ def create_option_parser(results)
                 "(Optional, defaults to 100)") do |concurrency|
           results[:concurrency] = concurrency
         end
+        opts.on('-t', '--connect-timeout TIMEOUT', Integer,
```
Comment (adreyer) [inline lib/bolt/cli.rb]: perhaps we shouldn't use -t for this

File: lib/bolt/config.rb [hunk 6]
```diff
@@ -20,9 +20,10 @@ module Bolt
       log_destination: STDERR
     }.freeze
 
-    TRANSPORT_OPTIONS = %i[insecure password run_as sudo sudo_password key tty user].freeze
+    TRANSPORT_OPTIONS = %i[insecure password run_as sudo sudo_password key tty user connect_timeout].freeze
```
Comment (adreyer) [inline lib/bolt/config.rb]: Should we load this from the config file too?

**Comment by nicklewis** [comment]
Updated to remove the short flag `-t` and to allow `connect-timeout` to be set in the transport-specific sections of the config.

**Comment by nicklewis** [comment]
Okay, tests are finally green. Hopefully they stay that way. I added some more helpful messaging around connection timeouts.

**Pull request merged**